test/hotspot/jtreg/gc/TestAlwaysPreTouchBehavior.java

```diff
@@ -28,103 +28,153 @@
  */
 
 /**
  * @test id=G1
  * @summary tests AlwaysPreTouch
  * @requires vm.gc.G1
- * @requires os.maxMemory > 2G
- * @requires os.family != "aix"
+ * @requires os.maxMemory > 2G & vm.flagless & os.family != "aix"
  * @library /test/lib
  * @build jdk.test.whitebox.WhiteBox
  * @run driver jdk.test.lib.helpers.ClassFileInstaller jdk.test.whitebox.WhiteBox
- * @run main/othervm -Xbootclasspath/a:. -XX:+UnlockDiagnosticVMOptions -XX:+WhiteBoxAPI -Xmx512m -Xms512m -XX:+AlwaysPreTouch gc.TestAlwaysPreTouchBehavior
+ * @run main/othervm -Xbootclasspath/a:. -XX:+UnlockDiagnosticVMOptions -XX:+WhiteBoxAPI -Xmx64m gc.TestAlwaysPreTouchBehavior -XX:+UseG1GC
  */
 
 /**
  * @test id=ZGenerational
  * @summary tests AlwaysPreTouch
  * @requires vm.gc.ZGenerational
- * @requires os.maxMemory > 2G
- * @requires os.family != "aix"
+ * @requires os.maxMemory > 2G & vm.flagless & os.family != "aix"
  * @library /test/lib
  * @build jdk.test.whitebox.WhiteBox
  * @run driver jdk.test.lib.helpers.ClassFileInstaller jdk.test.whitebox.WhiteBox
- * @run main/othervm -Xbootclasspath/a:. -XX:+UnlockDiagnosticVMOptions -XX:+WhiteBoxAPI -XX:+UseZGC -XX:+ZGenerational -Xmx512m -Xms512m -XX:+AlwaysPreTouch gc.TestAlwaysPreTouchBehavior
+ * @run main/othervm -Xbootclasspath/a:. -XX:+UnlockDiagnosticVMOptions -XX:+WhiteBoxAPI -Xmx64m gc.TestAlwaysPreTouchBehavior -XX:+UseZGC -XX:+ZGenerational
  */
 
 /**
  * @test id=ZSinglegen
  * @summary tests AlwaysPreTouch
  * @requires vm.gc.ZSinglegen
- * @requires os.maxMemory > 2G
- * @requires os.family != "aix"
+ * @requires os.maxMemory > 2G & vm.flagless & os.family != "aix"
  * @library /test/lib
  * @build jdk.test.whitebox.WhiteBox
  * @run driver jdk.test.lib.helpers.ClassFileInstaller jdk.test.whitebox.WhiteBox
- * @run main/othervm -Xbootclasspath/a:. -XX:+UnlockDiagnosticVMOptions -XX:+WhiteBoxAPI -XX:+UseZGC -XX:-ZGenerational -Xmx512m -Xms512m -XX:+AlwaysPreTouch gc.TestAlwaysPreTouchBehavior
+ * @run main/othervm -Xbootclasspath/a:. -XX:+UnlockDiagnosticVMOptions -XX:+WhiteBoxAPI -Xmx64m gc.TestAlwaysPreTouchBehavior -XX:+UseZGC -XX:-ZGenerational
  */
 
 /**
  * @test id=Epsilon
  * @summary tests AlwaysPreTouch
  * @requires vm.gc.Epsilon
- * @requires os.maxMemory > 2G
- * @requires os.family != "aix"
+ * @requires os.maxMemory > 2G & vm.flagless & os.family != "aix"
  * @library /test/lib
  * @build jdk.test.whitebox.WhiteBox
  * @run driver jdk.test.lib.helpers.ClassFileInstaller jdk.test.whitebox.WhiteBox
- * @run main/othervm -Xbootclasspath/a:. -XX:+UnlockDiagnosticVMOptions -XX:+WhiteBoxAPI -XX:+UnlockExperimentalVMOptions -XX:+UseEpsilonGC -Xmx512m -Xms512m -XX:+AlwaysPreTouch gc.TestAlwaysPreTouchBehavior
+ * @run main/othervm -Xbootclasspath/a:. -XX:+UnlockDiagnosticVMOptions -XX:+WhiteBoxAPI -Xmx64m gc.TestAlwaysPreTouchBehavior -XX:+UnlockExperimentalVMOptions -XX:+UseEpsilonGC
```

Comment:
Why the explicit `-Xmx64m`? As I understand this is essentially the launcher, whose heap-size is of little importance.
Also, why does the launch require `WhiteBoxAPI`?

Reply:
> Why the explicit -Xmx64m? As I understand this is essentially the launcher, whose heap-size is of little importance.

No particular reason, just don't like launchers to use large heaps. I can remove it.

> Also, why does the launch require WhiteBoxAPI?

Because the launcher needs to access hostAvailableMemory in order to decide before starting the test whether it makes sense to start the test.

Reply:
> just don't like launchers to use large heaps.

Could you add a comment at the start of this file explaining the test setup, launcher creating another VM + real test flags? There, the rational for small heap (64M) can be covered as well.

Some text on these fields can also help understand this test.
```
    final static long expectedMaxNonHeapRSS = M * 256;
    final static  long requiredAvailableBefore = heapsize * 2 + expectedMaxNonHeapRSS;
    final static  long requiredAvailableDuring = expectedMaxNonHeapRSS;
```

Reply:
Sure.